File: lib/booking-communication/schedule.ts
export function hoursBefore(iso: string, hours: number): Date {
  return new Date(new Date(iso).getTime() - hours * 60 * 60 * 1000);
}

/** If the computed send time is already past, send on the next cron tick. */
export function clampToNow(date: Date): Date {
  const now = Date.now();
  return date.getTime() < now ? new Date(now) : date;
}

export function h48SendAt(scheduledAtIso: string): Date {
  return clampToNow(hoursBefore(scheduledAtIso, 48));
}

export function h24SendAt(scheduledAtIso: string): Date {
  return clampToNow(hoursBefore(scheduledAtIso, 24));
}

export function h20SendAt(scheduledAtIso: string): Date {
  return clampToNow(hoursBefore(scheduledAtIso, 20));
}

const PARIS_TZ = "Europe/Paris";

type ParisDateParts = {
  year: number;
  month: number;
  day: number;
  weekday: string;
};

function getParisDateParts(instant: Date): ParisDateParts {
  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone: PARIS_TZ,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    weekday: "short",
  });
  const parts = formatter.formatToParts(instant);
  const lookup = Object.fromEntries(parts.map((part) => [part.type, part.value]));
  return {
    year: Number(lookup.year),
    month: Number(lookup.month),
    day: Number(lookup.day),
    weekday: lookup.weekday ?? "",
  };
}

function isWeekendWeekday(weekday: string): boolean {
  return weekday === "Sat" || weekday === "Sun";
}

function parisDateKey(parts: ParisDateParts): string {
  return `${parts.year}-${String(parts.month).padStart(2, "0")}-${String(parts.day).padStart(2, "0")}`;
}

function addParisCalendarDays(dateKey: string, deltaDays: number): string {
  const [year, month, day] = dateKey.split("-").map(Number);
  const utc = new Date(Date.UTC(year, month - 1, day + deltaDays, 12, 0, 0));
  return parisDateKey(getParisDateParts(utc));
}

/** 08:00 Europe/Paris on the given YYYY-MM-DD calendar day. */
export function parisAt8am(dateKey: string): Date {
  const [year, month, day] = dateKey.split("-").map(Number);
  const utcGuess = new Date(Date.UTC(year, month - 1, day, 8, 0, 0));
  const offsetFormatter = new Intl.DateTimeFormat("en-US", {
    timeZone: PARIS_TZ,
    timeZoneName: "shortOffset",
  });
  const offsetPart = offsetFormatter
    .formatToParts(utcGuess)
    .find((part) => part.type === "timeZoneName")?.value;
  const match = offsetPart?.match(/GMT([+-]\d{1,2})(?::(\d{2}))?/);
  const offsetHours = match ? Number(match[1]) : 1;
  const offsetMinutes = match?.[2] ? Number(match[2]) : 0;
  const totalOffsetMinutes = offsetHours * 60 + Math.sign(offsetHours) * offsetMinutes;
  return new Date(
    Date.UTC(year, month - 1, day, 8, 0, 0) - totalOffsetMinutes * 60 * 1000,
  );
}

/** Latest weekday 08:00 Paris that is still on or before `raw`. */
export function snapToPreviousWeekday8amParis(raw: Date): Date {
  let dateKey = parisDateKey(getParisDateParts(raw));
  let candidate = parisAt8am(dateKey);

  while (candidate.getTime() > raw.getTime()) {
    dateKey = addParisCalendarDays(dateKey, -1);
    candidate = parisAt8am(dateKey);
  }

  let weekday = getParisDateParts(candidate).weekday;
  while (isWeekendWeekday(weekday)) {
    dateKey = addParisCalendarDays(dateKey, -1);
    candidate = parisAt8am(dateKey);
    weekday = getParisDateParts(candidate).weekday;
  }

  return candidate;
}

export function roleSeq48SendAt(scheduledAtIso: string): Date {
  return clampToNow(snapToPreviousWeekday8amParis(hoursBefore(scheduledAtIso, 48)));
}

export function roleSeq24SendAt(scheduledAtIso: string): Date {
  return clampToNow(snapToPreviousWeekday8amParis(hoursBefore(scheduledAtIso, 24)));
}

const ROLE_RECOVERY_COMPRESSED_GAP_MS = 5 * 60 * 1000;
const ROLE_RECOVERY_MONDAY_GAP_MS = 5 * 60 * 1000;

export type ParisWeekdayShort =
  | "Mon"
  | "Tue"
  | "Wed"
  | "Thu"
  | "Fri"
  | "Sat"
  | "Sun";

export type RecoveryWeekdayVariant =
  | "monday_meeting"
  | "tuesday_meeting"
  | "wednesday_meeting";

/** Weekday of the meeting in Europe/Paris (Mon–Sun). */
export function meetingWeekdayParis(scheduledAtIso: string): ParisWeekdayShort {
  return getParisDateParts(new Date(scheduledAtIso)).weekday as ParisWeekdayShort;
}

/** Latest occurrence of `targetWeekday` at 08:00 Paris on or before `before`. */
export function previousWeekday8amParis(
  targetWeekday: ParisWeekdayShort,
  before: Date,
): Date {
  let dateKey = parisDateKey(getParisDateParts(before));

  for (let i = 0; i < 14; i++) {
    const candidate = parisAt8am(dateKey);
    const weekday = getParisDateParts(candidate).weekday;
    if (weekday === targetWeekday && candidate.getTime() <= before.getTime()) {
      return candidate;
    }
    dateKey = addParisCalendarDays(dateKey, -1);
  }

  throw new Error(
    `No ${targetWeekday} 08:00 Paris on or before ${before.toISOString()}`,
  );
}

/** Latest Saturday 08:00 Paris on or before `before`. */
export function previousSaturday8amParis(before: Date): Date {
  return previousWeekday8amParis("Sat", before);
}

/**
 * Recovery schedule for Mon/Tue/Wed meetings (Europe/Paris weekday).
 * Does not clamp to now — cron sends when scheduled_for is due.
 */
export function planRecoveryByMeetingWeekday(
  scheduledAtIso: string,
): {
  roleSeq48: Date;
  roleSeq24: Date;
  variant: RecoveryWeekdayVariant;
} {
  const meeting = new Date(scheduledAtIso);
  const weekday = meetingWeekdayParis(scheduledAtIso);

  if (weekday === "Mon") {
    const roleSeq48 = previousSaturday8amParis(meeting);
    const roleSeq24 = new Date(roleSeq48.getTime() + ROLE_RECOVERY_MONDAY_GAP_MS);
    return { roleSeq48, roleSeq24, variant: "monday_meeting" };
  }

  if (weekday === "Tue") {
    return {
      roleSeq48: previousSaturday8amParis(meeting),
      roleSeq24: previousWeekday8amParis("Mon", meeting),
      variant: "tuesday_meeting",
    };
  }

  if (weekday === "Wed") {
    return {
      roleSeq48: previousWeekday8amParis("Mon", meeting),
      roleSeq24: previousWeekday8amParis("Tue", meeting),
      variant: "wednesday_meeting",
    };
  }

  throw new Error(
    `planRecoveryByMeetingWeekday: unsupported meeting weekday ${weekday}`,
  );
}

/** True when the meeting is less than 48h away (normal snap window missed). */
export function isRoleRecoveryCompressed(scheduledAtIso: string): boolean {
  return hoursBefore(scheduledAtIso, 48).getTime() <= Date.now();
}

export function planRoleRecoverySchedule(scheduledAtIso: string): {
  roleSeq48: Date;
  roleSeq24: Date;
  compressed: boolean;
} {
  if (isRoleRecoveryCompressed(scheduledAtIso)) {
    const roleSeq48 = clampToNow(new Date());
    const roleSeq24 = new Date(roleSeq48.getTime() + ROLE_RECOVERY_COMPRESSED_GAP_MS);
    return { roleSeq48, roleSeq24, compressed: true };
  }

  return {
    roleSeq48: roleSeq48SendAt(scheduledAtIso),
    roleSeq24: roleSeq24SendAt(scheduledAtIso),
    compressed: false,
  };
}

File: app/streamlit_booking_resend/schedule.py
"""Port of lib/booking-communication/schedule.ts for UI preview."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import TypedDict
from zoneinfo import ZoneInfo

PARIS_TZ = ZoneInfo("Europe/Paris")
ROLE_RECOVERY_COMPRESSED_GAP = timedelta(minutes=5)
ROLE_RECOVERY_MONDAY_GAP = timedelta(minutes=5)


class RoleRecoverySchedule(TypedDict):
    role_seq_48: datetime
    role_seq_24: datetime
    compressed: bool


class RecoveryWeekdaySchedule(TypedDict):
    role_seq_48: datetime
    role_seq_24: datetime
    variant: str


class MainSchedule(TypedDict):
    immediate: datetime
    h48_confirm: datetime | None
    h24_relance: datetime | None
    h20_cancel: datetime | None


def _parse_iso(iso: str) -> datetime:
    normalized = iso.replace("Z", "+00:00")
    dt = datetime.fromisoformat(normalized)
    if dt.tzinfo is None:
        return dt.replace(tzinfo=UTC)
    return dt


def hours_before(iso: str, hours: int) -> datetime:
    start = _parse_iso(iso)
    return start - timedelta(hours=hours)


def clamp_to_now(date: datetime) -> datetime:
    now = datetime.now(UTC)
    if date.tzinfo is None:
        date = date.replace(tzinfo=UTC)
    return now if date < now else date


def h48_send_at(scheduled_at_iso: str) -> datetime:
    return clamp_to_now(hours_before(scheduled_at_iso, 48))


def h24_send_at(scheduled_at_iso: str) -> datetime:
    return clamp_to_now(hours_before(scheduled_at_iso, 24))


def h20_send_at(scheduled_at_iso: str) -> datetime:
    return clamp_to_now(hours_before(scheduled_at_iso, 20))


def _paris_date_key(instant: datetime) -> str:
    paris = instant.astimezone(PARIS_TZ)
    return f"{paris.year:04d}-{paris.month:02d}-{paris.day:02d}"


def _paris_weekday(instant: datetime) -> str:
    return instant.astimezone(PARIS_TZ).strftime("%a")


def meeting_weekday_paris(scheduled_at_iso: str) -> str:
    return _paris_weekday(_parse_iso(scheduled_at_iso))


def previous_weekday_8am_paris(target_weekday: str, before: datetime) -> datetime:
    if before.tzinfo is None:
        before = before.replace(tzinfo=UTC)
    date_key = _paris_date_key(before)

    for _ in range(14):
        candidate = paris_at_8am(date_key)
        weekday = _paris_weekday(candidate)
        if weekday == target_weekday and candidate.timestamp() <= before.timestamp():
            return candidate
        date_key = _add_paris_calendar_days(date_key, -1)

    raise ValueError(
        f"No {target_weekday} 08:00 Paris on or before {before.isoformat()}",
    )


def previous_saturday_8am_paris(before: datetime) -> datetime:
    return previous_weekday_8am_paris("Sat", before)


def plan_recovery_by_meeting_weekday(scheduled_at_iso: str) -> RecoveryWeekdaySchedule:
    meeting = _parse_iso(scheduled_at_iso)
    weekday = meeting_weekday_paris(scheduled_at_iso)

    if weekday == "Mon":
        role_seq_48 = previous_saturday_8am_paris(meeting)
        role_seq_24 = role_seq_48 + ROLE_RECOVERY_MONDAY_GAP
        return {
            "role_seq_48": role_seq_48,
            "role_seq_24": role_seq_24,
            "variant": "monday_meeting",
        }

    if weekday == "Tue":
        return {
            "role_seq_48": previous_saturday_8am_paris(meeting),
            "role_seq_24": previous_weekday_8am_paris("Mon", meeting),
            "variant": "tuesday_meeting",
        }

    if weekday == "Wed":
        return {
            "role_seq_48": previous_weekday_8am_paris("Mon", meeting),
            "role_seq_24": previous_weekday_8am_paris("Tue", meeting),
            "variant": "wednesday_meeting",
        }

    raise ValueError(
        f"plan_recovery_by_meeting_weekday: unsupported meeting weekday {weekday}",
    )


def _is_weekend_weekday(weekday: str) -> bool:
    return weekday in {"Sat", "Sun"}


def _add_paris_calendar_days(date_key: str, delta_days: int) -> str:
    year, month, day = map(int, date_key.split("-"))
    utc = datetime(year, month, day, 12, 0, 0, tzinfo=UTC) + timedelta(days=delta_days)
    return _paris_date_key(utc)


def paris_at_8am(date_key: str) -> datetime:
    year, month, day = map(int, date_key.split("-"))
    return datetime(year, month, day, 8, 0, 0, tzinfo=PARIS_TZ)


def snap_to_previous_weekday_8am_paris(raw: datetime) -> datetime:
    if raw.tzinfo is None:
        raw = raw.replace(tzinfo=UTC)
    date_key = _paris_date_key(raw)
    candidate = paris_at_8am(date_key)

    while candidate > raw.astimezone(PARIS_TZ):
        date_key = _add_paris_calendar_days(date_key, -1)
        candidate = paris_at_8am(date_key)

    weekday = _paris_weekday(candidate)
    while _is_weekend_weekday(weekday):
        date_key = _add_paris_calendar_days(date_key, -1)
        candidate = paris_at_8am(date_key)
        weekday = _paris_weekday(candidate)

    return candidate


def role_seq_48_send_at(scheduled_at_iso: str) -> datetime:
    raw = hours_before(scheduled_at_iso, 48)
    return clamp_to_now(snap_to_previous_weekday_8am_paris(raw))


def role_seq_24_send_at(scheduled_at_iso: str) -> datetime:
    raw = hours_before(scheduled_at_iso, 24)
    return clamp_to_now(snap_to_previous_weekday_8am_paris(raw))


def is_role_recovery_compressed(scheduled_at_iso: str) -> bool:
    return hours_before(scheduled_at_iso, 48) <= datetime.now(UTC)


def plan_role_recovery_schedule(scheduled_at_iso: str) -> RoleRecoverySchedule:
    if is_role_recovery_compressed(scheduled_at_iso):
        role_seq_48 = clamp_to_now(datetime.now(UTC))
        role_seq_24 = role_seq_48 + ROLE_RECOVERY_COMPRESSED_GAP
        return {
            "role_seq_48": role_seq_48,
            "role_seq_24": role_seq_24,
            "compressed": True,
        }

    return {
        "role_seq_48": role_seq_48_send_at(scheduled_at_iso),
        "role_seq_24": role_seq_24_send_at(scheduled_at_iso),
        "compressed": False,
    }


def plan_main_schedule(
    scheduled_at_iso: str,
    *,
    category: str = "agence",
    sequence_starts_at: datetime | None = None,
) -> MainSchedule:
    immediate = clamp_to_now(sequence_starts_at or datetime.now(UTC))
    if not scheduled_at_iso:
        return {
            "immediate": immediate,
            "h48_confirm": None,
            "h24_relance": None,
            "h20_cancel": None,
        }
    h48 = h48_send_at(scheduled_at_iso)
    h24 = h24_send_at(scheduled_at_iso)
    if category != "agence":
        return {
            "immediate": immediate,
            "h48_confirm": h48,
            "h24_relance": h24,
            "h20_cancel": None,
        }
    return {
        "immediate": immediate,
        "h48_confirm": h48,
        "h24_relance": h24,
        "h20_cancel": h20_send_at(scheduled_at_iso),
    }


def format_paris(dt: datetime | None) -> str:
    if dt is None:
        return "—"
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=UTC)
    return dt.astimezone(PARIS_TZ).strftime("%Y-%m-%d %H:%M")
